cli/bench/lsp.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::error::generic_error;
+use deno_core::error::AnyError;
+use deno_core::serde::de;
+use deno_core::serde::Deserialize;
+use deno_core::serde_json;
+use deno_core::serde_json::json;
+use deno_core::serde_json::Value;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::process::Stdio;
+use tokio::io::AsyncBufReadExt;
+use tokio::io::AsyncReadExt;
+use tokio::io::AsyncWriteExt;
+use tokio::io::BufReader;
+use tokio::io::BufWriter;
+use tokio::io::Lines;
+use tokio::process::ChildStderr;
+use tokio::process::ChildStdin;
+use tokio::process::ChildStdout;
+use tokio::process::Command;
+use tokio::runtime;
+use tokio::time::Instant;
+
+lazy_static! {
+  static ref CONTENT_TYPE_REG: Regex =
+    Regex::new(r"(?i)^content-length:\s+(\d+)").unwrap();
+}
+
+#[derive(Debug, Deserialize)]
+struct LspResponse {
+  pub result: Option<Value>,
+  pub id: u32,
+  pub error: Option<LspResponseError>,
+}
+
+#[derive(Debug, Deserialize)]
+struct LspResponseError {
+  pub code: i32,
+  pub message: String,
+  pub data: Option<Value>,
+}
+
+#[derive(Debug, Deserialize)]
+struct LspNotification {
+  pub method: String,
+  pub params: Option<Value>,
+}
+
+fn create_basic_runtime() -> runtime::Runtime {
+  runtime::Builder::new_current_thread()
+    .enable_io()
+    .enable_time()
+    // for performance purposes, we might as well allow this to go to as
+    // many threads as it can, so that the harness doesn't become the bottle
+    // neck
+    // .max_blocking_threads(32)
+    .build()
+    .unwrap()
+}
+
+pub(crate) fn run_local<F, R>(future: F) -> R
+where
+  F: std::future::Future<Output = R>,
+{
+  let rt = create_basic_runtime();
+  rt.block_on(future)
+}
+
+struct LspClient {
+  pub output: Lines<BufReader<ChildStderr>>,
+  reader: BufReader<ChildStdout>,
+  start: Instant,
+  writer: BufWriter<ChildStdin>,
+}
+
+impl LspClient {
+  pub fn new(deno_exe: &PathBuf) -> Result<Self, AnyError> {
+    let mut child = Command::new(deno_exe)
+      .kill_on_drop(true)
+      .args(&["lsp"])
+      .stderr(Stdio::piped())
+      .stdin(Stdio::piped())
+      .stdout(Stdio::piped())
+      .spawn()?;
+
+    let stdout = child
+      .stdout
+      .take()
+      .ok_or_else(|| generic_error("could not get handle to stdout"))?;
+    let reader = BufReader::new(stdout);
+
+    let stdin = child
+      .stdin
+      .take()
+      .ok_or_else(|| generic_error("could not get handle to stdin"))?;
+    let writer = BufWriter::new(stdin);
+
+    let stderr = child
+      .stderr
+      .take()
+      .ok_or_else(|| generic_error("could not get handle to stderr"))?;
+    let output = BufReader::new(stderr).lines();
```

Comment:
Is data transmitted over stderr? Maybe it's easier to debug if stderr inherits

Reply:
No data is transmitted... it is just what the lsp logs, like:

```
Starting Deno language server...
  version: 1.7.4 (debug, x86_64-apple-darwin)
Connected to "Visual Studio Code" 1.53.2
language server initialized
Server ready.
```

Reply:
In the case of a panic, it would be nice to see that in the logs.
I suggest removing the `stderr(Stdio::piped())` bit.

Reply:
inherit, especially for benchmarking, will fill stderr with a lot of useless startup information... as discussed we should pipe it to null